### PR TITLE
perf: Switched to `Vec` + created benchmarks

### PR DESCRIPTION
**Pull request by Norbiros**
This PR makes reading `complex_player.dat` around 50% faster, which makes our library faster than FastNBT.

### CODE REVIEW

File: src/nbt/compound.rs [hunk 1]
```diff
@@ -3,23 +3,23 @@ use bytes::{Buf, BufMut, Bytes, BytesMut};
...
 pub struct NbtCompound {
-    pub child_tags: HashMap<String, NbtTag>,
+    pub child_tags: Vec<(String, NbtTag)>,
 }
 
 impl NbtCompound {
     pub fn new() -> NbtCompound {
         NbtCompound {
-            child_tags: HashMap::new(),
+            child_tags: Vec::new(),
```
Comment (goteusz-maszyk) [inline src/nbt/compound.rs:17]:
```suggestion
            child_tags: vec![],
```

File: src/nbt/compound.rs [hunk 2]
```diff
@@ -30,7 +30,7 @@ impl NbtCompound {
             let name = get_nbt_string(bytes)?;
 
             if let Ok(tag) = NbtTag::deserialize_data(bytes, tag_id) {
-                child_tags.insert(name, tag);
+                child_tags.push((name, tag));
```
Comment (goteusz-maszyk) [inline src/nbt/compound.rs:33]: Potentially duplicated keys

**Review by goteusz-maszyk — CHANGES_REQUESTED** [review]

File: src/nbt/compound.rs (hunk 1 above)
Comment (Norbiros) [reply src/nbt/compound.rs:17]: Nope. We used a lot of vec new in our codebase

File: benches/read.rs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        input = buffer;
+    }
+
+    let bytes = Bytes::from_iter(input);
+    let bytes_mut = BytesMut::from(bytes);
+
+    let mut group = c.benchmark_group("read");
+    group.throughput(Throughput::Bytes(bytes_mut.len() as u64));
+
+    group.bench_function("read_bigtest_nbt", |b| {
```
Comment (SzczurekYT) [inline benches/read.rs:26]: Here it is called bigtest, there it is complex_player. One name would be better,

File: benches/read.rs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        b.iter(|| {
+            let output = Nbt::read(&mut bytes_mut.clone()).expect("Failed to parse NBT");
+            black_box(output)
```
Comment (SzczurekYT) [inline benches/read.rs:29]: I must check something here, will do that tomorrow.

**Review by SzczurekYT — CHANGES_REQUESTED** [review]
Split this up into two commits, first one to add benchmarks and then second one that does the change, so we can compare the performance difference, and so there is nice history.

File: benches/read.rs (hunk 4 above)
Comment (SzczurekYT) [reply benches/read.rs:29]: Yup, I was right
That is the proper way to bench that function
```rust
 b.iter_batched_ref(
            || bytes_mut.clone(),
            |bytes| Nbt::read(bytes).expect("Failed to parse NBT"),
            BatchSize::SmallInput,
        )
```
That bytes.clone skewed the benchmark by around 20 MiB/s

File: benches/read.rs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    let mut buffer = Vec::new();
+    file.read_to_end(&mut buffer).expect("Failed to read file");
+    let mut src = &buffer[..];
+
+    let mut src_decoder = GzDecoder::new(&mut src);
+    let mut input = Vec::new();
+    if src_decoder.read_to_end(&mut input).is_err() {
+        input = buffer;
+    }
+
```
Comment (SzczurekYT) [inline benches/read.rs:19]: Extract that to a separate function.

**Review by SzczurekYT — CHANGES_REQUESTED** [review]

File: benches/read.rs (hunk 4 above)
Comment (Norbiros) [reply benches/read.rs:29]: Hmm. Bytes clone wasn't supposed to be instant?

File: Cargo.toml [hunk 6]
```diff
@@ -23,3 +23,15 @@ cesu8 = "1.1.0"
 derive_more = { version = "1.0.0", features = ["into", "from"] }
 thiserror = "1.0.63"
 serde = { version = "1.0.209", optional = true, features = ["derive"] }
+rustc-hash = "2.0.0"
```
Comment (SzczurekYT) [inline Cargo.toml:26]: Unused dependency.

File: src/macros.rs [hunk 7]
```diff
@@ -64,9 +64,14 @@ macro_rules! nbt_inner {
...
-            $(map.insert($key.into(), nbt_inner!(@value $value));)*
-            map
+            let mut values = ::std::vec::Vec::<(::std::string::String, $crate::NbtTag)>::new();
+            $(
+                let key: ::std::string::String = $key.into();
+                if !values.iter().any(|(k, _)| k == &key) {
+                    values.push((key, nbt_inner!(@value $value)));
+                }
+            )*
+            values
```
Comment (SzczurekYT) [inline src/macros.rs:74]: That duplicate checking should be in the from iter implementation. not here

**Review by SzczurekYT — CHANGES_REQUESTED** [review]

**Comment by SzczurekYT** [comment]
Alright, extract benchmark addition to separate commit (make it first), and squash everything else into second one, and you can go.

**Pull request merged**